### PR TITLE
[GSoC] Add new Hashtable based TSDF volume to RGBD module

### PR DESCRIPTION
**Pull request by akashsharma02**
This work is part of GSoC and is potentially for the first evaluation phase of the program. 

My mentor for the GSoC time period is: @savuor
My proposal is available here: https://summerofcode.withgoogle.com/dashboard/project/6190777371197440/details/

Please note that this is still Work in Progress, as the system is not yet reliable and does not work well. 
Making an initial PR as requested for early code review. 

### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch

### CODE REVIEW

File: modules/rgbd/src/hash_tsdf.cpp [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+                while (tcurr < tmax)
+                {
+                    Point3f currRayPos          = orig + tcurr * rayDirV;
+                    cv::Vec3i currVolumeUnitIdx = volume.volumeToVolumeUnitIdx(currRayPos);
+
+                    VolumeUnitMap::const_iterator it = volume.volumeUnits.find(currVolumeUnitIdx);
+
+                    TsdfType currTsdf = prevTsdf;
+                    int currWeight    = 0;
+                    float stepSize    = blockSize;
```
Comment (savuor) [inline modules/rgbd/src/hash_tsdf.cpp:565]: Such big step can skip a subvolume. The line between `currRayPos` and `currRayPos + blockSize * rayDirV` can enter new subvolume near its corner and then exit it near the same corner.

File: modules/rgbd/src/kinfu.cpp [hunk 2]
```diff
@@ -60,7 +61,8 @@ Ptr<Params> Params::defaultParams()
 
     // default pose of volume cube
     p.volumePose = Affine3f().translate(Vec3f(-volSize/2.f, -volSize/2.f, 0.5f));
-    p.tsdf_trunc_dist = 0.04f; //meters;
+    /* p.volumePose  = Affine3f(); */
+    p.tsdf_trunc_dist = 5 * p.voxelSize; //meters;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp:65]: The pose change is an experimental thing. However, in general, it is better to have the truncation distance as a multiple to voxel size, as opposed to a constant, because with different resolutions (coarse or fine) the integration might become problematic when truncation distance goes below voxel size.

File: modules/rgbd/src/hash_tsdf.cpp [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+                    if (accessVolUnits.emplace(tsdf_idx).second)
+                    {
+                        VolumeUnit volumeUnit;
+                        cv::Point3i volumeDims(volume.volumeUnitResolution,
+                                               volume.volumeUnitResolution,
+                                               volume.volumeUnitResolution);
+
+                        cv::Affine3f subvolumePose =
+                            volume.pose.translate(volume.volumeUnitIdxToVolume(tsdf_idx));
+                        volumeUnit.pVolume = cv::makePtr<TSDFVolumeCPU>(
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:109]: The bug was here at line [103](https://github.com/opencv/opencv_contrib/pull/2566/files/70b1dfbd3ebbc99391a8dced6192139bd7d84e91#diff-f658111e4b84275e8f54acba97231a01R103) 
Using a new accessVolUnits set for every integration meant that, previously added tsdf indices would be overwritten. [](url)

File: modules/rgbd/src/hash_tsdf.cpp (hunk 1 above)
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:565]: Setting the step to 0.5 * blockSize fixes all the raycasting glitches.

File: modules/rgbd/src/hash_tsdf.cpp [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+                        //! TODO: Figure out voxel interpolation
+                        Voxel currVoxel = currVolumeUnit->at(volUnitLocalIdx);
+                        currTsdf        = currVoxel.tsdf;
+                        /* currTsdf   = volume.interpolateVoxel(currRayPos); */
+                        /* std::cout << "Current TSDF: " << currTsdf << " interpolated: " << interpolatedTsdf << "\n"; */
+                        currWeight = currVoxel.weight;
+                        stepSize   = tstep;
+                    }
+                    //! Surface crossing
+                    if (prevTsdf > 0.f && currTsdf <= 0.f && currWeight > 0)
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:452]: Done.

File: modules/rgbd/src/hash_tsdf.cpp [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+        {
+            cv::Vec3i tsdf_idx         = totalVolUnits[i];
+            VolumeUnitMap::iterator it = volume.volumeUnits.find(tsdf_idx);
+            if (it == volume.volumeUnits.end())
+                return;
+
+            Point3f volumeUnitPos     = volume.volumeUnitIdxToVolume(it->first);
+            Point3f volUnitInCamSpace = vol2cam * volumeUnitPos;
+            if (volUnitInCamSpace.z < rgbd::Odometry::DEFAULT_MIN_DEPTH() ||
+                volUnitInCamSpace.z > rgbd::Odometry::DEFAULT_MAX_DEPTH())
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp]: Done.

File: modules/rgbd/src/kinfu.cpp [hunk 6]
```diff
@@ -120,7 +124,7 @@ class KinFuImpl : public KinFu
     Params params;
 
     cv::Ptr<ICP> icp;
-    cv::Ptr<TSDFVolume> volume;
+    cv::Ptr<HashTSDFVolume> volume;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp]: Done. There is an issue with running with TSDFVolume, will fix eventually.

File: modules/rgbd/samples/kinfu_demo.cpp [hunk 7]
```diff
@@ -26,39 +27,39 @@ static vector<string> readDepth(std::string fileList)
     vector<string> v;
 
     fstream file(fileList);
-    if(!file.is_open())
+    if (!file.is_open())
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/samples/kinfu_demo.cpp:30]: Done.

File: modules/rgbd/src/tsdf.cpp [hunk 8]
```diff
@@ -316,83 +305,89 @@ struct IntegrateInvoker : ParallelLoopBody
...
+            for (int y = 0; y < volume.volResolution.y; y++)
             {
-                Voxel* volDataY = volDataX+y*volume.volDims[1];
-                // optimization of camSpace transformation (vector addition instead of matmul at each z)
-                Point3f basePt = vol2cam*(Point3f(x, y, 0)*volume.voxelSize);
+                TsdfVoxel* volDataY = volDataX + y * volume.volDims[1];
+                // optimization of camSpace transformation (vector addition instead of matmul at
+                // each z)
+                Point3f basePt = vol2cam * (Point3f(x, y, 0) * volume.voxelSize);
+                /* std::cout << "Base Point: " << basePt << " TsdfVoxelSize: " << volume.voxelSize
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:361]: the same

File: modules/rgbd/src/tsdf.cpp [hunk 9]
```diff
@@ -629,87 +633,86 @@ inline Point3f TSDFVolumeCPU::getNormalVoxel(Point3f p) const
...
+          // We do subtract voxel size to minimize checks after
+          // Note: origin of volume coordinate is placed
+          // in the center of voxel (0,0,0), not in the corner of the voxel!
+          boxMax(volume.volSize - Point3f(volume.voxelSize, volume.voxelSize, volume.voxelSize)),
+          boxMin(),
+          cam2vol(volume.pose.inv() * cameraPose),
+          vol2cam(cameraPose.inv() * volume.pose),
+          reproj(intrinsics.makeReprojector())
+    {
+        std::cout << "Entered RaycastInvoker:" << std::endl;
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:682]: the same

File: modules/rgbd/src/tsdf.cpp [hunk 10]
```diff
@@ -822,57 +828,61 @@ struct RaycastInvoker : ParallelLoopBody
...
     {
+        std::cout << " Entered raycastInvoker" << std::endl;
         const Point3f camTrans = cam2vol.translation();
-        const Matx33f  camRot  = cam2vol.rotation();
-        const Matx33f  volRot  = vol2cam.rotation();
-
-        for(int y = range.start; y < range.end; y++)
+        const Matx33f camRot = cam2vol.rotation();
+        const Matx33f volRot = vol2cam.rotation();
+        std::cout << "range.start: " << range.start << " range.end: " << range.end << "\n";
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:837]: the same

File: modules/rgbd/src/tsdf.cpp [hunk 11]
```diff
@@ -822,57 +828,61 @@ struct RaycastInvoker : ParallelLoopBody
...
                 float tmin = max(max(max(minAx.x, minAx.y), max(minAx.x, minAx.z)), clip);
-                float tmax =     min(min(maxAx.x, maxAx.y), min(maxAx.x, maxAx.z));
+                float tmax = min(min(maxAx.x, maxAx.y), min(maxAx.x, maxAx.z));
 
                 // precautions against getting coordinates out of bounds
                 tmin = tmin + tstep;
                 tmax = tmax - tstep;
-
-                if(tmin < tmax)
+                std::cout << "Tmin: " << tmin << "tmax: " << tmax << "\n";
```
Comment (savuor) [inline modules/rgbd/src/tsdf.cpp:871]: the same

File: modules/rgbd/src/kinfu.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp:66]: Done. Changed to 7 * voxel_size for 512 resolution and 2* voxel_size for 128 resolution

File: modules/rgbd/src/kinfu.cpp [hunk 12]
```diff
@@ -86,8 +88,9 @@ Ptr<Params> Params::coarseParams()
     float volSize = 3.f;
     p->volumeDims = Vec3i::all(128); //number of voxels
     p->voxelSize  = volSize/128.f;
+    p->tsdf_trunc_dist = 5 * p->voxelSize;
 
-    p->raycast_step_factor = 0.75f;  //in voxel sizes
+    p->raycast_step_factor = 0.25f;  //in voxel sizes
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp]: I have changed the values back to the original, so not required.

File: modules/rgbd/src/tsdf.cpp [hunk 13]
```diff
@@ -2,67 +2,26 @@
 // It is subject to the license terms in the LICENSE file found in the top-level directory
 // of this distribution and at http://opencv.org/license.html
 
-// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this module's directory
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/tsdf.cpp:5]: Done..

File: modules/rgbd/include/opencv2/rgbd/kinfu.hpp [hunk 14]
```diff
@@ -126,7 +138,7 @@ struct CV_EXPORTS_W Params
 
     How much voxel sizes we skip each raycast step
     */
-    CV_PROP_RW float raycast_step_factor;
+    CV_PROP_RW float raycastStepFactor;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/kinfu.hpp:141]: Done.

File: modules/rgbd/samples/kinfu_demo.cpp [hunk 15]
```diff
@@ -371,32 +362,35 @@ int main(int argc, char **argv)
...
-    //params->volumePose = Affine3f().translate(Vec3f(-cubeSize/2.f, -cubeSize/2.f, 0.25f)); //meters
-    //params->tsdf_max_weight = 16;
-
-    if(!idle)
+    // float cubeSize = 1.f;
+    // params->voxelSize = cubeSize/params->volumeDims[0]; //meters
+    // params->tsdf_trunc_dist = 0.01f; //meters
+    // params->icpDistThresh = 0.01f; //meters
+    // params->volumePose = Affine3f().translate(Vec3f(-cubeSize/2.f, -cubeSize/2.f, 0.25f));
+    // //meters params->tsdf_max_weight = 16;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/samples/kinfu_demo.cpp:391]: Done.

File: modules/rgbd/src/kinfu.cpp [hunk 16]
```diff
@@ -238,6 +257,7 @@ bool KinFuImpl<T>::updateT(const T& _depth)
             return false;
 
         pose = pose * affine;
+        /* std::cout << "Obtained pose:\n " << pose.matrix << "\n"; */
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp:260]: Done.

File: modules/rgbd/src/kinfu.cpp [hunk 17]
```diff
@@ -272,36 +291,38 @@ void KinFuImpl<T>::render(OutputArray image, const Matx44f& _cameraPose) const
...
         volume->raycast(cameraPose, params.intr, params.frameSize, points, normals);
         renderPointsNormals(points, normals, image, params.lightPose);
     }
 }
 
 
-template< typename T >
-void KinFuImpl<T>::getCloud(OutputArray p, OutputArray n) const
+template< typename MatType >
+void KinFuImpl<MatType>::getCloud(cv::OutputArray p, OutputArray n) const
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu.cpp:308]: Done.

File: modules/rgbd/src/kinfu_frame.cpp [hunk 18]
```diff
@@ -624,13 +624,16 @@ void makeFrameFromDepth(InputArray _depth,
...
-        threshold(depth, depth, truncateThreshold, 0.0, THRESH_TOZERO_INV);
+        threshold(smooth, depthThreshold, truncateThreshold * depthFactor, 0.0, THRESH_TOZERO_INV);
+    else
+        depthThreshold = smooth;
 
     // we don't need depth pyramid outside this method
     // if we do, the code is to be refactored
 
-    Depth scaled = smooth;
+    Depth scaled = depthThreshold;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/kinfu_frame.cpp:636]: Done.

File: modules/rgbd/src/volume.hpp [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+// TODO: Optimization possible:
+// * TsdfType can be FP16
+// * weight can be uint16
+typedef float TsdfType;
+struct TsdfVoxel
+{
+    TsdfType tsdf;
+    int weight;
+};
+typedef Vec<uchar, sizeof(TsdfVoxel)> VecTsdfVoxel;
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/volume.hpp:55]: Makes sense. I was thinking we could have all structures say (EsdfVoxel) in future in a single file, but this is also alright.

File: modules/rgbd/src/tsdf.cpp [hunk 20]
```diff
@@ -2,81 +2,35 @@
...
+      maxWeight(_maxWeight)
 {
     // Unlike original code, this should work with any volume size
     // Not only when (x,y,z % 32) == 0
-
-    volSize = Point3f(volResolution) * voxelSize;
-
-    truncDist = std::max(_truncDist, 2.1f * voxelSize);
+    volSize   = Point3f(volResolution) * voxelSize;
+    truncDist = std::max(_truncDist, 4.0f * voxelSize);
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/tsdf.cpp:26]: Considering that this class is still not exposed: Actually this affects HashTSDFVolume since the HashTSDF uses TSDFVolume units. So, it would be better to keep this as 4.0f, although in **my opinion**, this is not required at all.

File: modules/rgbd/src/tsdf.cpp [hunk 21]
```diff
@@ -117,148 +67,184 @@ void TSDFVolumeCPU::reset()
...
+{
+    //! Out of bounds
+    if ((volumeIdx[0] >= volResolution.x || volumeIdx[0] < 0) ||
+        (volumeIdx[1] >= volResolution.y || volumeIdx[1] < 0) ||
+        (volumeIdx[2] >= volResolution.z || volumeIdx[2] < 0))
+    {
+        TsdfVoxel dummy;
+        dummy.tsdf   = 1.0f;
+        dummy.weight = 0;
+        std::cout << "Returning dummy\n";
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/tsdf.cpp:87]: Done.

File: modules/rgbd/src/tsdf.cpp [hunk 22]
```diff
@@ -117,148 +67,184 @@ void TSDFVolumeCPU::reset()
...
             {
-                Voxel* volDataY = volDataX + y*volume.volDims[1];
-                // optimization of camSpace transformation (vector addition instead of matmul at each z)
-                Point3f basePt = vol2cam*(Point3f((float)x, (float)y, 0)*volume.voxelSize);
+                TsdfVoxel* volDataY = volDataX + y * volume.volDims[1];
+                // optimization of camSpace transformation (vector addition instead of matmul at
+                // each z)
+                Point3f basePt = vol2cam * (Point3f((float)x, (float)y, 0) * volume.voxelSize);
                 v_float32x4 camSpacePt(basePt.x, basePt.y, basePt.z, 0);
+                /* std::cout << "basePt: " << basePt << " zstep: " << zStepPt << "\n"; */
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/tsdf.cpp:225]: Done.

File: modules/rgbd/include/opencv2/rgbd/kinfu.hpp [hunk 23]
```diff
@@ -210,13 +222,13 @@ class CV_EXPORTS_W KinFu
 
         @param points vector of points which are 4-float vectors
      */
-    CV_WRAP virtual void getPoints(OutputArray points) const = 0;
+    /* CV_WRAP virtual void getPoints(OutputArray points) const = 0; */
```
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/kinfu.hpp]: Done, but not tested. Will write as part of tests.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (savuor) [reply modules/rgbd/src/hash_tsdf.cpp:453]: This is not resolved yet and there is a problem:
If surface crossing happened and `tInterp` is not a good number (which is possible if `prevTsdf` is close to `currTsdf`: this leads to Inf or even NaN if `stepSize` is also small so that `tcurr` is close to `tprev`)
OR
If the voxels are OK but their neighbours are not good so that it's impossible to get normal
THEN
We skip this step and search for the next point on a ray.
BUT we shouldn't.
A loop should be finished if we faced a surface, even if we cannot get feasible points/normals from it (in that case we should indicate it with NaNs).

[the remark this answers is not in the document]
Comment (savuor) [reply modules/rgbd/src/hash_tsdf.cpp:453]: What should be done is to run loop until surface crossing, mark it with a bool, after the loop check that bool and if it's true then perform the interpolation, normal calculation and so on, if it's false then write NaNs to output pixels.

File: modules/rgbd/src/hash_tsdf.cpp [hunk 24]
```diff
@@ -0,0 +1,661 @@
...
+    if (_points.needed())
+    {
+        std::vector<std::vector<ptype>> pVecs, nVecs;
+
+        std::vector<Vec3i> totalVolUnits;
+        for (const auto& keyvalue : volumeUnits)
+        {
+            totalVolUnits.push_back(keyvalue.first);
+        }
+        std::cout << "Number of volumeUnits in volume(fetchPoints): " << totalVolUnits.size()
```
Comment (savuor) [inline modules/rgbd/src/hash_tsdf.cpp:587]: extra print

File: modules/rgbd/src/hash_tsdf.cpp [hunk 25]
```diff
@@ -0,0 +1,661 @@
...
+                float tmin = 0;
+                float tmax = volume.truncateThreshold / rayDirV.z;
+                float tcurr = tmin;
+
+                cv::Vec3i prevVolumeUnitIdx =
+                    cv::Vec3i(std::numeric_limits<int>::min(), std::numeric_limits<int>::min(),
+                              std::numeric_limits<int>::min());
+
+                float tprev = tcurr;
+                TsdfType prevTsdf = float16_t(volume.truncDist);
```
Comment (savuor) [inline modules/rgbd/src/hash_tsdf.cpp:423]: Looks like you did some experiments with `float16` but forgot to delete it from here.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:453]: I had fixed this by moving the break outside the (is tInterp invalid check), but it seems it was not pushed correctly for some reason. I think just breaking the loop regardless of the tInterp value at a surface crossing solves the issue.

**Comment by savuor** [comment]
👍

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp [hunk 26]
```diff
@@ -0,0 +1,76 @@
+#ifndef __OPENCV_RGBD_INTRINSICS_HPP__
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:1]: Short license header is required.

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+#define __OPENCV_RGBD_INTRINSICS_HPP__
+
+#include "opencv2/core/matx.hpp"
+
+namespace cv
+{
+    namespace kinfu
+    {
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:9]: Avoid indentation in namespaces.

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+struct Intr
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:10]: > Intr

Use meaningful names instead.

File: modules/rgbd/include/opencv2/rgbd/volume.hpp [hunk 29]
```diff
@@ -0,0 +1,58 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this
+// module's directory
+
+#ifndef __OPENCV_VOLUME_H__
+#define __OPENCV_VOLUME_H__
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/volume.hpp:9]: > __OPENCV_VOLUME_H__

Missing module name: `__OPENCV_RGBD_VOLUME_H__`

File: modules/rgbd/include/opencv2/rgbd/volume.hpp [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace cv
+{
+namespace kinfu
+{
+
+class Volume
+{
+   public:
+    explicit Volume(float _voxelSize, cv::Affine3f _pose, float _raycastStepFactor)
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/volume.hpp:22]: > explicit 

Makes sense for cases with single parameter only.

File: modules/rgbd/include/opencv2/rgbd/volume.hpp [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+          pose(_pose),
+          raycastStepFactor(_raycastStepFactor)
+    {
+    }
+
+    virtual ~Volume(){};
+
+    virtual void integrate(InputArray _depth, float depthFactor, cv::Affine3f cameraPose,
+                           cv::kinfu::Intr intrinsics)                                     = 0;
+    virtual void raycast(cv::Affine3f cameraPose, cv::kinfu::Intr intrinsics, cv::Size frameSize,
```
Comment (alalek) [inline modules/rgbd/include/opencv2/rgbd/volume.hpp:34]: "const reference" for non-trivial types.

File: modules/rgbd/samples/kinfu_demo.cpp [hunk 32]
```diff
@@ -2,13 +2,14 @@
 // It is subject to the license terms in the LICENSE file found in the top-level directory
 // of this distribution and at http://opencv.org/license.html
 
-// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this module's directory
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this
+// module's directory
```
Comment (alalek) [inline modules/rgbd/samples/kinfu_demo.cpp:6]: avoid unrelated changes

File: modules/rgbd/samples/kinfu_demo.cpp [hunk 33]
```diff
@@ -2,13 +2,14 @@
...
 
-#include <iostream>
 #include <fstream>
-#include <opencv2/imgproc.hpp>
+#include <iostream>
 #include <opencv2/calib3d.hpp>
 #include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
 #include <opencv2/rgbd/kinfu.hpp>
```
Comment (alalek) [inline modules/rgbd/samples/kinfu_demo.cpp:13]: avoid unnecessary changes

File: modules/rgbd/samples/kinfu_demo.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/rgbd/samples/kinfu_demo.cpp:30]: not done (please point on commit with non-style changes in this file)

File: modules/rgbd/src/hash_tsdf.cpp [hunk 34]
```diff
@@ -0,0 +1,627 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this
+// module's directory
```
Comment (alalek) [inline modules/rgbd/src/hash_tsdf.cpp:6]: perhaps not related to this functionality

File: modules/rgbd/src/hash_tsdf.cpp [hunk 35]
```diff
@@ -0,0 +1,627 @@
...
+#include "hash_tsdf.hpp"
```
Comment (alalek) [inline modules/rgbd/src/hash_tsdf.cpp:7]: `precomp.hpp` should be the first include of src/*.cpp files

File: modules/rgbd/src/hash_tsdf.cpp [hunk 36]
```diff
@@ -0,0 +1,627 @@
...
+    HashTSDFVolumeCPU& volume;
+    const Depth& depth;
+    const Intr::Reprojector reproj;
+    const cv::Affine3f cam2vol;
+    const float depthFactor;
+    const int depthStride;
+    mutable Mutex mutex;
+};
+
+struct VolumeUnitInFrustumInvoker : ParallelLoopBody
```
Comment (alalek) [inline modules/rgbd/src/hash_tsdf.cpp:131]: consider using of lambda functions instead for all `ParallelLoopBody` stuff

File: modules/rgbd/src/hash_tsdf.hpp [hunk 37]
```diff
@@ -0,0 +1,114 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this
+// module's directory
```
Comment (alalek) [inline modules/rgbd/src/hash_tsdf.hpp:6]: note

File: modules/rgbd/src/hash_tsdf.cpp (hunk 36 above)
Comment (savuor) [reply modules/rgbd/src/hash_tsdf.cpp:131]: Do we prefer lambdas over functor objects just because of coding style or `ParallelLoopBody` misses some important features?

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp (hunk 28 above)
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:10]: This was existing code that was moved to a different file and will cause a change in variable name in a lot of files.

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp (hunk 27 above)
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:9]: Done.

File: modules/rgbd/include/opencv2/rgbd/intrinsics.hpp (hunk 26 above)
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/intrinsics.hpp:1]: Done.

File: modules/rgbd/include/opencv2/rgbd/volume.hpp (hunk 29 above)
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/volume.hpp:9]: Done.

File: modules/rgbd/include/opencv2/rgbd/volume.hpp (hunk 30 above)
Comment (akashsharma02) [reply modules/rgbd/include/opencv2/rgbd/volume.hpp:22]: Done.

File: modules/rgbd/samples/kinfu_demo.cpp (hunk 33 above)
Comment (akashsharma02) [reply modules/rgbd/samples/kinfu_demo.cpp:13]: Reverted all the style changes in this file.

File: modules/rgbd/samples/kinfu_demo.cpp (hunk 32 above)
Comment (akashsharma02) [reply modules/rgbd/samples/kinfu_demo.cpp:6]: Done.

File: modules/rgbd/samples/kinfu_demo.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (akashsharma02) [reply modules/rgbd/samples/kinfu_demo.cpp:30]: Reverted all the style changes in the file. Only added useHashTSDF parameter to the demo

File: modules/rgbd/src/hash_tsdf.cpp (hunk 35 above)
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:7]: Done. but this file currently does not have any code optimization.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 36 above)
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp]: I have changed a few small `parallel_for_` to lambdas.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 35 above)
Comment (alalek) [reply modules/rgbd/src/hash_tsdf.cpp:4]: .cpp files only, not .hpp files.

File: modules/rgbd/src/hash_tsdf.hpp [hunk 38]
```diff
@@ -0,0 +1,112 @@
...
+    float truncDist;
+    float truncateThreshold;
+    int volumeUnitResolution;
+    float volumeUnitSize;
+    bool zFirstMemOrder;
+};
+
+struct VolumeUnit
+{
+    explicit VolumeUnit() : pVolume(nullptr){};
```
Comment (alalek) [inline modules/rgbd/src/hash_tsdf.hpp:40]: `explicit` is misused.

File: modules/rgbd/src/tsdf.hpp [hunk 39]
```diff
@@ -2,51 +2,111 @@
 // It is subject to the license terms in the LICENSE file found in the top-level directory
 // of this distribution and at http://opencv.org/license.html
 
-// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this module's directory
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this
+// module's directory
```
Comment (alalek) [inline modules/rgbd/src/tsdf.hpp:6]: unrelated change

File: modules/rgbd/src/utils.hpp [hunk 40]
```diff
@@ -9,6 +9,8 @@
 #ifndef __OPENCV_RGBD_UTILS_HPP__
 #define __OPENCV_RGBD_UTILS_HPP__
 
+#include "precomp.hpp"
```
Comment (alalek) [inline modules/rgbd/src/utils.hpp:12]: remove `precomp.hpp` from .hpp files.

File: modules/rgbd/src/volume.cpp [hunk 41]
```diff
@@ -0,0 +1,29 @@
+#include <opencv2/rgbd/volume.hpp>
```
Comment (alalek) [inline modules/rgbd/src/volume.cpp:1]: - license
- precomp.hpp

File: modules/rgbd/src/hash_tsdf.hpp (hunk 38 above)
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.hpp:40]: Done.

File: modules/rgbd/src/tsdf.hpp (hunk 39 above)
Comment (akashsharma02) [reply modules/rgbd/src/tsdf.hpp:6]: done.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 35 above)
Comment (akashsharma02) [reply modules/rgbd/src/hash_tsdf.cpp:4]: Okay, done. Is this a preference, or is it that, every header which has optimizations declared requires this?

File: modules/rgbd/src/utils.hpp (hunk 40 above)
Comment (akashsharma02) [reply modules/rgbd/src/utils.hpp:12]: done.

File: modules/rgbd/src/volume.cpp (hunk 41 above)
Comment (akashsharma02) [reply modules/rgbd/src/volume.cpp:1]: done.

File: modules/rgbd/src/hash_tsdf.cpp (hunk 35 above)
Comment (alalek) [reply modules/rgbd/src/hash_tsdf.cpp:5]: It is some kind of this: https://en.wikipedia.org/wiki/Precompiled_header

**Comment by akashsharma02** [comment]
Hi @alalek, I have fixed all the review comments, can the PR be merged? Or is there something missing from my side?
